lite/provider.go

```diff
@@ -30,3 +33,105 @@ type PersistentProvider interface {
 	// SaveFullCommit saves a FullCommit (without verification).
 	SaveFullCommit(fc FullCommit) error
 }
+
+// A provider that can update itself w/ more recent commit data.
+type UpdatingProvider interface {
+	Provider
+
+	// Update internal information by fetching information somehow.
+	// UpdateToHeight will block until the request is complete,
+	// or returns an error if the request cannot complete.
+	// Generally, one must call UpdateToHeight(h) before LatestFullCommit(_,h,h)
+	// will return this height.
+	//
+	// NOTE: Behavior with concurrent requests is undefined.  To make
+	// concurrent calls safe, look at the struct `ConcurrentUpdatingProvider`.
+	UpdateToHeight(chainID string, height int64) error
+}
+
+//----------------------------------------
+
+type concurrentProvider struct {
+	UpdatingProvider
+
+	// pending map to synchronize concurrent verification requests
+	mtx                  sync.Mutex
+	pendingVerifications map[pendingKey]*pendingResult
+}
+
+// convenience to create the key for the lookup map
+type pendingKey struct {
+	chainID string
+	height  int64
+}
+
+// used to cache the result from underlying UpdatingProvider.
+type pendingResult struct {
+	wait chan struct{}
+	err  error // cached result.
+}
+
+func NewConcurrentUpdatingProvider(up UpdatingProvider) concurrentProvider {
+	return concurrentProvider{
+		UpdatingProvider:     up,
+		pendingVerifications: make(map[pendingKey]*pendingResult),
+	}
+}
+
+// Returns the unique pending request for all identical calls to
+// joinConcurrency(chainID,height), and returns true for isFirstCall only for the
+// first call, which should call the returned callback w/ results if any.
+// The callback must be called, otherwise there will be memory leaks.
+// Other subsequent calls should just return uniq.err.
+// NOTE: This is a separate function, primarily to make mtx unlocking more obviously safe via defer.
+func (cp concurrentProvider) joinConcurrency(chainID string, height int64) (uniq *pendingResult, isFirstCall bool, callback func(error)) {
+	cp.mtx.Lock()
+	defer cp.mtx.Unlock()
+
+	var pkey = pendingKey{chainID, height}
+	cp.mtx.Lock()
+	if uniq = cp.pendingVerifications[pkey]; uniq != nil {
+		cp.mtx.Unlock()
+		<-uniq.wait // uniq.wait is of type `chan struct{}`
+		return uniq, false, nil
+	} else {
+		uniq = &pendingResult{wait: make(chan struct{}), err: nil}
+		cp.pendingVerifications[pkey] = uniq
+		cp.mtx.Unlock()
```

Comment:
Why the additional `cp.mtx.Unlock()`s everywhere? Doesn't the `defer cp.mtx.Unlock()` above be sufficient?

Reply:
It's a callback (not original function)